pkg/cli/cachedaemon.go

```diff
@@ -0,0 +1,169 @@
+				config = zap.NewDevelopmentConfig()
+			}
+
+			config.Encoding = encoding
+			config.Level = zap.NewAtomicLevelAt(*level)
+			config.EncoderConfig.EncodeTime = zapcore.ISO8601TimeEncoder
+
+			err = logger.Init(config)
+			if err != nil {
+				return fmt.Errorf("could not initialize logger: %w", err)
+			}
+
+			ctx := cmd.Context()
+
+			if profilePath != "" {
+				file, err := os.Create(profilePath)
+				if err != nil {
+					return fmt.Errorf("cannot open profile path %s: %w", profilePath, err)
+				}
+				_ = pprof.StartCPUProfile(file)
+				profilerEnabled = true
+			}
+
+			if tracing {
+				shutdownTelemetry = telemetry.Init(ctx, telemetry.Server)
+			}
+
+			cert, err := tls.LoadX509KeyPair(certFile, keyFile)
+			if err != nil {
+				return fmt.Errorf("cannot open TLS cert and key files (%s, %s): %w", certFile, keyFile, err)
+			}
+
+			cl, err := client.NewClient(ctx, upstreamHost, upstreamPort, client.WithheadlessHost(headlessHost))
+			if err != nil {
+				return err
+			}
+			ctx = client.IntoContext(ctx, cl)
+
+			listen, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+			if err != nil {
+				return fmt.Errorf("failed to listen on TCP port %d: %w", port, err)
+			}
+
+			s := cached.NewServer(ctx, cl, &cert, stagingPath)
+
+			osSignals := make(chan os.Signal, 1)
+			signal.Notify(osSignals, os.Interrupt, syscall.SIGTERM)
+			go func() {
+				<-osSignals
+				s.Grpc.GracefulStop()
+			}()
+
+			err = s.Cached.Prepare(ctx)
+			if err != nil {
+				return fmt.Errorf("failed to prepare cache daemon in %s: %w", stagingPath, err)
+			}
+
+			logger.Info(ctx, "start server", key.Port.Field(port), key.Environment.Field(env.String()))
+			return s.Serve(listen)
+		},
+		PostRunE: func(cmd *cobra.Command, _ []string) error {
+			if shutdownTelemetry != nil {
+				shutdownTelemetry()
+			}
+
+			if profilerEnabled {
+				pprof.StopCPUProfile()
+			}
+
+			return nil
+		},
+	}
+
+	flags := cmd.PersistentFlags()
+
+	level = zap.LevelFlag("log-level", zap.DebugLevel, "Log level")
+	flags.AddGoFlag(flag.CommandLine.Lookup("log-level"))
+	flags.StringVar(&encoding, "log-encoding", "console", "Log encoding (console | json)")
+	flags.BoolVar(&tracing, "tracing", false, "Whether tracing is enabled")
+	flags.StringVar(&profilePath, "profile", "", "CPU profile output path (profiling enabled if set)")
+	flags.StringVar(&certFile, "cert", "development/server.crt", "TLS cert file")
+	flags.StringVar(&keyFile, "key", "development/server.key", "TLS key file")
+	flags.StringVar(&upstreamHost, "upstream-host", "localhost", "GRPC server hostname")
+	flags.Uint16Var(&upstreamPort, "upstream-port", 5051, "GRPC server port")
+	flags.StringVar(&headlessHost, "headless-host", "", "Alternative headless hostname to use for round robin connections")
+	flags.UintVar(&timeout, "timeout", 0, "GRPC client timeout (ms)")
+	flags.IntVar(&port, "port", 5053, "cache API port")
+	flags.StringVar(&stagingPath, "staging-path", "", "path for staging downloaded caches")
```

Comment:
This is where `cached` will put the downloaded cache, and what has to be on the same device as the paths it then later links the cache to.